### DISCUSSION TITLE
Versioning object on relationship update causes duplicated UPDATE statements which fail

### DISCUSSION THREAD

**Discussion by jsormaz** [sqlalchemy/sqlalchemy#10798] [Usage Questions] [answered]
I'm going to try to put together a minimal reproduction, but in the meantime wanted to get a discussion going and see if anyone has ideas.

My use case:

I want to increment the version_id column of a parent object when a child is added/removed from a one to many on many to many on the parent. 

I setup a hook on the relationships:

```python
@event.listens_for(Parent.children, "append")
@event.listens_for(Parent.children, "remove")
def receive_append(target, value, initiator):
     if inspect(target).persistent:
          flag_modified(target, 'id') # could be any column on the parent object, just flag it as modified to force an update and trigger version_id_generator
```

My thinking here is to just flag an attribute as modified to force an UPDATE statement.

However, when I do this, it actually produces 2 UPDATE statements.

The first one will update 'id' and 'version_id' as expected.

The second UPDATE statement then attempts to increment 'version_id' again, but fails to match a row because it is SELECTing against the old version_id

Its worth noting that I have overriden all my relationships with the following:

```python
from functools import partial
relationship = partial(relationship, lazy="selectin", join_depth=1)
```

This is kind of needed for performance reasons in my application. I have a lot of deeply connected models, and I want to load the first layer of relationships any time I access them. But maybe I'm not really understanding the downstream effects.

I'm thinking that maybe because child also has a relationship, (child.parent), that it is also loading the parent into the session and causing the double update? But it doesn't really make sense to me, shouldn't that get taken care of by the session identity manager?

**Comment by zzzeek** [comment]
shrugs, i would need to see the point of origin for the two UPDATE statements, if it is doing a SELECT that suggests an expired or unloaded attribute is being accessed at an unusual time but overall a simple "flag_modified" would not have any effects like this

**Comment by jsormaz** [comment]
I managed to get a reproduction.
Of course, this actually helped me find the root cause. I still don't really understand it, but maybe you have some insight

In my app I am tracking an `updated_by_id` column on all my models. I had to use `post_update=True` to break the circular dependency.
When I was picking a column name I know all my models have to use in `flag_modified()`, I happened to pick `updated_by_id`
If I change it just to use `id`, then my problem seems to go away.

I'm still interested to get some feedback if you think this is an actual issue, or just a caveat of the use case.

BTW, thanks so much for all you do on this project. 

```python
from sqlalchemy import (
    Column,
    inspect,
    create_engine,
    event,
    ForeignKey,
    Integer,
)
from sqlalchemy.orm import (
    declarative_base,
    declared_attr,
    attributes,
    relationship,
    Session,
    sessionmaker,
)

from functools import partial
relationship = partial(relationship, lazy="selectin", join_depth=1)

Base = declarative_base()
engine = create_engine("sqlite://", echo=True)
Session = sessionmaker(engine)


class BaseMixin(Base):
    __abstract__ = True

    @declared_attr
    def id(cls):
        return Column(Integer, primary_key=True)

    @declared_attr
    def updated_by_id(cls):
        return Column(
            Integer,
            ForeignKey("user.id"),
            default=1,
            onupdate=1,
        )

    @declared_attr
    def updated_by(cls):
        return relationship(
            "User",
            foreign_keys=[cls.updated_by_id],
            post_update=True,
        )

    @declared_attr
    def version_id(cls):
        return Column(Integer)

    @declared_attr
    def __mapper_args__(cls):
        return {
            "version_id_col": cls.version_id,
        }

class User(BaseMixin):
    __tablename__ = "user"


class Parent(BaseMixin):
    __tablename__ = "parent"

    children = relationship("Child", back_populates="parent")


@event.listens_for(Parent.children, "append")
def receive_append(target, value, initiator):
    if inspect(target).persistent:
        print(f"Flagging {target} as modified")
        attributes.flag_modified(target, "updated_by_id")


class Child(BaseMixin):
    __tablename__ = "child"

    parent_id = Column(Integer, ForeignKey("parent.id"))
    parent = relationship(Parent, back_populates="children")


Base.metadata.create_all(engine)
session = Session()
u1 = User(id=1)
p1 = Parent(id=1, updated_by=u1)
c1 = Child(id=1)  # create a child but don't add it to the parent yet
session.add(u1)
session.add(p1)
session.add(c1)

session.commit()

try:
    c1_transient = Child(id=1)
    p1_transient = Parent(
        id=1,
        version_id=p1.version_id,
        children=[c1_transient],
    )

    session.merge(p1_transient)
    session.commit()

except Exception as e:
    print(str(e))

```

**Reply by zzzeek** [reply, marked as answer]
well  you have "post_update" set on the `updated_by` relationship so that's the source of your second UPDATE statement.  the "post update" steps run in all cases, and their logic is 1. iterate through all objects that were modified 2. look for changes to the columns that are singificant for the post update, that is, the columns which are managed by the `relationship()`, in this case `updated_by_id` 3. emit an UPDATE statement for each object where a change was registered for that element.

What is happening here is that the merge() step is creating a change artifact for the version_id_col that is interfering with the ORM unit of work process from making use of the newly updated version_id value for the second UPDATE statement being emitted.  so this is a bug. The bug is either that a. the version_id col is being bumped twice in one unit of work flush or b. when the two bumps happen, the second one is not reliably getting the correct value.   I think a fix for now would have to target b. so #10800 is added for that.

**Reply by jsormaz** [reply]
Wow, thank you for the detailed explanation, this really helps my understanding.

**Reply by zzzeek** [reply]
well it's a bug so I had to figure it out exactly :/
